### ISSUE TITLE
Anti-pattern question

### ISSUE THREAD

**Issue by 0x6e6562** [diegohaz/constate#45]
Is it an anti-pattern to cache the `setContextState` function for subsequent use by a non-React-Context component?

For example, this approach appears to work in practice:

```js

// Non-context module

export const onMount = ({ setContextState }) => {
  cachedSetContextState = setContextState;
};
```
Which can be registered with a `<Provider>`:

```jsx
<Provider onMount={onMount} />
```
So that later on in the app the cached context state can be accessed to update the context state by a component that isn't otherwise bound into the React context:

```js

// In the same non-context module as above
// Has access to the cached `cachedSetContextState` function
cachedSetContextState('context', { foo: 'bar' });
```

Is this a bad idea?

What I think I trying to achieve is something like [channels](https://github.com/redux-saga/redux-saga/blob/master/docs/advanced/Channels.md) in `redux-saga`, i.e. to be able to mutate the state from a non-GUI source.

**Comment by diegohaz** [comment]
Hey, Ben.

The only downside I can think of is that it doesn't work with server side rendering (if that's a thing for you). But there may be others.

I'd love to see an experiment on it though. Keep me posted if you take it forward. 😄

**Comment by diegohaz** [comment]
Will leave this open so others can chime in.

**Comment by 0x6e6562** [comment]
Maybe I misunderstood what you said earlier.

I now think I've got it to work with SSR, but I needed to create the context in the rendering pipeline (using Gatsby this achieved with a [replaceRenderer](https://www.gatsbyjs.org/docs/ssr-apis/#replaceRenderer) callback that configures the SSR process, but other frameworks would presumably have a similar pipeline approach).

Or did I completely miss the point?

**Comment by diegohaz** [comment]
I was saying that it wouldn't work with SSR because you were using `onMount`, which is called inside `Provider`'s `componentDidMount`, which in turn isn't called on server (unless Gatsby has a different mechanism to handle it).

I'm not sure if I understand how `replaceRenderer` works. If you could put a codesandbox example together, I think it would be really useful for people interested on this (myself included 😄). They now support Gatsby sandboxes.

**Comment by 0x6e6562** [comment]
Good idea - [here's an example](https://codesandbox.io/s/62m5ky8w0k) that shows how this works with Gatsby.

The good news is that codesandbox can run `gatsby develop` and `gatsby build`.

The bad news is that I don't think they run `gatsby serve` (so that you see what the SSR behavior would be). I've played around with trying to get this to deploy to `now.sh` from `codesandbox`, but for some reason the deployment is hardcoded to serve from the wrong directory and I can't see how to configure this.

So what I did was to download the example from codesandbox as a zip and then

```bash
yarn install && gatsby build && gatsby serve
```

and the result is the same as `gatsby develop`.

Long story, short point - `onMount` from `constate` appears to behave the same way whether it is booted from `gatsby-browser.js` or `gatsby-ssr.js`.

So that's what I was thinking `onMount` would work for SSR.
